File: Pieces/queen.py
from Pieces.piece import Piece
from Pieces.moveCandidate import MoveCandidate
from enums import Colour
from coords import Coords

class Queen(Piece):
    def __init__(self, colour: Colour) -> None:
        super().__init__(colour)
        self.value = 9

    def get_representation(self) -> str:
        return 'q' if self.colour == Colour.BLACK else 'Q'

    def get_candidate_moves(self, coords: Coords):
        # list of tuples of new coordinates the piece can go
        candidates = []

        for di, dj in [
            [-1, -1], [-1, 1], [1, -1], [1, 1],
            [-1, 0], [1, 0], [0, -1], [0, 1]
            ]:

            candidate = MoveCandidate(di, dj, 8)
            candidates.append(candidate)
        
        return candidates

File: rust_chess/src/pieces/queen.rs
use crate::enums::Colour;
use crate::coords::Coords;
use crate::moves::move_ray::{self, MoveRay};
use crate::pieces::Piece;

#[derive(Debug, Clone)]
pub struct Queen {
    pub colour: Colour,
}

impl Queen {
    pub fn new(colour: Colour) -> Self {
        Self { colour }
    }
}

impl Piece for Queen {
    fn colour(&self) -> Colour {
        self.colour
    }

    fn get_representation(&self) -> char {
        match self.colour {
            Colour::White => 'Q',
            Colour::Black => 'q',
        }
    }

    fn get_destination_coords(&self, from: Coords) -> Vec<MoveRay> {
        let mut move_rays = Vec::new();

        for (di, dj) in &[
            (-1, -1), (-1, 0), (-1, 1),
            ( 0, -1)         , ( 0, 1),
            ( 1, -1), ( 1, 0), ( 1, 1)
        ] {
            move_rays.push(MoveRay {
                rank_diff: *di,
                file_diff: *dj,
                dist: 7,
                capture_allowed: true,
                capture_forced: false,
            });
        }
    }
}